CBC repeated IV

AleX04Nov opened the pull request:
This PR adds the `CBC repeated IV` demo.

AleX04Nov committed:
init commit

init commit

G1gg1L3s commented on cbc-repeated-iv/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+# need to use aes
+aes = "0.8.1"
+
+# to return structure `anyhow::Result<()>`
```
Cool, but there is no need of documenting every dependency, at least I haven't seen it in the wild :)
It can be useful, you want to specify something extraordinary, like a `TODO` or `FIXME`.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+            std::io::stdin().read_line(&mut line)?;
+            // remove new line char if present and leading and tracing spaces
+            line = line.trim().to_string();
+            match line.to_lowercase().as_str() {
```
These two lines contain two allocations: first in `to_string`, and second in `to_lowercase`.  It could be eliminated to only one:
```suggestion
            match line.trim().to_lowercase().as_str() {
```

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        let mut rng = rand::thread_rng();
+        self.key = rng.gen();
+        self.iv = rng.gen();
+    }
```
This method is used only once and during initialization. One of the rust's philosophy is "if the object exists, it should be already in a valid state". In this case, we can move these two lines into `new` method.

Just as example, imagine this as a piece of real crypto code. Making key initialization as a separate method will create possibilities for users to misuse this object and create weaknesses in systems. Moving key generation into `new` will eliminate such possibilities.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+            _ => (
+                "What are you doing here, step-bro??? UwU",
+                "Okay, here's some small secret crypto-tip for you: \
+                https://www.youtube.com/watch?v=dQw4w9WgXcQ",
```
<img src="https://user-images.githubusercontent.com/43041209/172223620-6710f4a3-e8fc-420c-9dd4-3446ff25c780.png " data-canonical-src="https://user-images.githubusercontent.com/43041209/172223620-6710f4a3-e8fc-420c-9dd4-3446ff25c780.png g" width="200" />

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+            example_desc_start,
+            example_desc_difference,
+            ascii_art
+        ) = match iter {
```
I would refactor this match statement into separate function, because the `example_eavesdropped_package` is huge and hard to read

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        Aes128CbcDec::new(&self.key.into(), &self.iv.into())
+            .decrypt_padded_vec_mut::<Pkcs7>(ciphertext_vec)
+            .ok()
+            .unwrap()
```
Why do you use `unwrap` here? Decryption is operation that by definition should fail in most of the cases. If you are sure that there is no way it could fail, explain it with comments or `expect`. Or just return `anyhow::Result<Vec<u8>>` and propagate error with `?` operator.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+                "Example with package that has the different name field being intercepted",
+                "Block #1 differs. We can assume that data in that block has changed. Also keeping \
+                in mind, that the first block is the same, so structure of json can be the same. \
+                That's why we can guess that usertype has changed and maybe blocks after that...",
```
As I see, almost all of this comments are the same and differ only in the block number. Maybe we can reuse them with fomat strings?

AleX04Nov replied:
Well, the first and the last one iteration have different comments. So I don't think that we need to make our code more complex for the sake of 2 iterations.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        for (block_num, block) in input_vec.chunks(block_size).enumerate() {
+            write!(
+                formatted_blocks,
+                "|            Block #{}            ",
```
You allow passing any block_size into a function, but headers are precalculated to use 16-byte headers :)
In this case, I would just create a global `BLOCK_SIZE` constant and use it everywhere and together with these headers.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        self.print_known_texts();
+
+        while iter < self.ciphertexts_vec.len() {
+            let mut line = String::new();
```
You could move this line out of the loop, and clear the buffer on every iteration. Just as a fancy and unnecessary optimization :)

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        println!("Leaked plaintext:");
+        println!("{formatted_blocks_pt}");
+        println!("{formatted_known_pt}");
+        println!("{}", " ".repeat(145).on_color(Color::Magenta));
```
Where did this number come from?

G1gg1L3s reviewed (commented):
Also, very important comment about PRs and readme:
The first thing in the PR is its description. Usually it contains a description of the changes, the goal of the PR (why it exists in the first place), technical details and how to run/test it. The exact content depends on a kind of PR, project and people you work with (there are people that will not look at the PR entirely, if there is no description), but it would be good to see at least some of these sections.

In our case, the same is true for readme as well. It's a demo, so we should explain at least how to run it and what it shows.

vixentael commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        let (formatted_blocks_ct, formatted_known_ct) =
+            CBCGame::format_blocks_hex(&self.known_ciphertext, 16);
+
+        println!("Getting corresponding known CT to leaked PT:");
+        println!("Here's leaked plaintext:");
+        println!("{formatted_blocks_pt}");
+        println!("{formatted_known_pt}\n");
+        println!("Here's corresponding CT to leaked PT:");
+        println!("{formatted_blocks_ct}");
+        println!("{formatted_known_ct}\n\n");
```
Let's not use "CT" and "PT" in prints, let's use "ciphertext" and "plaintext" instead.

vixentael commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,397 @@
+        println!("{formatted_blocks_ct}");
+        println!("{formatted_known_ct}\n\n");
+
+        println!("Let's wait and eavesdrop other CTs:");
```
same

vixentael commented:
+1 to readme.

i run the example, but what it does? i put "c" 4 times and the program exits. i see wild things in my console, so some descriptions would be nice.

AleX04Nov committed:
constant BLOCK_SIZE variable

Created constant BLOCK_SIZE in cbc_game file.
Using it instead of custom block size in `format_blocks` functions and KEY, IV array hardcoded `16` length.

AleX04Nov committed:
move string-match into separate function

Moving string-match from `CBCGame::example_eavesdropped_package` into `CBCGAME::get_desc_strings`

AleX04Nov committed:
move blocks_status into a separate function

move blocks_status into a separate function

AleX04Nov committed:
changing output. adding ascii tables

Changing the output of the program for every iteration. Now outputting beautified ASCII tables.

AleX04Nov committed:
removing commentaries from cargo

Removing commentaries from cargo file.

AleX04Nov committed:
updating `line` var in `CBCGame.start()`

Now, the `line` variable is being declared and initialised at the beginning of the `start` method.
Also moved `line` trimming into declaring of match statement.

AleX04Nov committed:
init random key and iv in `new` method

Now initialising key and init vector with random bytes in the `CBCGame::new` method.

AleX04Nov committed:
removing `unwrap` from `decrypt_text` method

Removing `unwrap` from `decrypt_text` method. Now this method returns `Result<Vec<u8>, UnpadError>`.
Also changed handling the return of `decrypt_text` method in `example_eavesdropped_package` method

AleX04Nov committed:
renaming CT and PT to CipherText and PlainText

Renaming CT and PT to CipherText and PlainText throughout the output of the program.

AleX04Nov committed:
added line about the ending of interactive

Added a fancy line in the `CBCGame::start` method to notify a user about the ending of interactive.

AleX04Nov committed:
create README

Creating README with small description.

G1gg1L3s commented on cbc-repeated-iv/README.md:
The _reusing init vector_ is repeated twice at the end. I would rephrase it to something like:
> Every launch of this demo generates unique key and init vector. But since the IV is reused every iteration, we can gain some information about the content of the messages.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,528 @@
+        strings_arr: &[&String],
+        strings_length: &[usize],
+        padding: usize,
+    ) -> anyhow::Result<String> {
```
This function returns result, but you use only `expect` here. It would be easier and better to use the `?` operator.

The `unwrap` and `expect` are generally used in places, where errors should not occur, but if the do, this is a bug, so we allow process to fail with ugly message and stacktrace.

But the errors that are part of a normal flow, like creating a file, generally are returned to the caller. In this way, a user (of a library for example) can use that error to print a nice message and recover the normal flow.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,528 @@
+            ],
+            3,
+        )
+        .expect("Ascii Table Creation Exception");
```
The same is here. The panic are not exceptions as in other languages. They are (with some omission) unrecoverable.

But if you want just to stop and exit from function in case of some error, just use `?`. It's really easy and convenient.

You can also look at the source code of some rust crates or even standard library, to get a feeling of error handling :)

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,528 @@
+            ],
+            &[
+                16,
+                String::from_utf8_lossy(&self.known_plaintext).len() + 19,
+                hex::encode(&self.known_ciphertext).len() + 20,
+                hex::encode(eavesdropped_packet).len() + 27,
```
Hard-coded constant are generally a bad sign. In this small demo they could be okay-ish. But, I would suggest you  to look and try out at some ascii table libraries. They, as I remember, support colors, utf8 chars (even emojis) and are generally easy to use.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,528 @@
+
+        println!("{}", self.cbc_same_iv_ascii_art.on_color(Color::Green));
+
+        self.reinitialize_enc_dec();
```
We already initialize the key and iv. Seems like a leftover :)

AleX04Nov replied:
I think, that we can leave it. Because it would show that when the interactive-game restarts - new keys are being generated. It's some kind of consistency, so the user can call the `start` method one more time and will remain with new keys and new iv.

G1gg1L3s replied:
If you want, sure it can. But in this case I think this method should call `encrypt_all_texts`, so the object will be in consistent state.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,528 @@
+
+        writeln!(result_string, "╔{}╗", "═".repeat(size)).expect("");
+        writeln!(result_string, "║{}║", " ".repeat(size)).expect("");
+        for string_index in 0..strings_arr.len() {
```
You could replace index-based loop with an iterator-based one
```suggestion
        for (string, length) in string_arr.iter().zip(strings_length) {
```

AleX04Nov committed:
change ascii table

Use `prettytable-rs` to form ASCII tables.
Now in `create_table` using `prettytable-rs` crate.

AleX04Nov committed:
update output. Changed terminology. Changed Order

Now only using 'known' and 'eavesdropped' terminology.
Now known texts would always be the first ones.

AleX04Nov committed:
updating readme

updating readme

AleX04Nov committed:
update error handlings

Updating handling errors in `CBCGame::create_table`, `CBCGame::example_eavesdropped_package` and `CBCGame::start` methods.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,575 @@
+        padding: usize,
+    ) -> anyhow::Result<String> {
+        // At first let's create and build custom styles for title and main table
+        let main_table_format = FormatBuilder::new()
+            .column_separator('║')
+            .borders('║')
+            .separators(&[LinePosition::Top], LineSeparator::new('═', '╦', '╔', '╗'))
+            .separators(
+                &[LinePosition::Intern],
+                LineSeparator::new('═', '╬', '╠', '╣'),
+            )
+            .separators(
+                &[LinePosition::Bottom],
+                LineSeparator::new('═', '╩', '╚', '╝'),
+            )
+            .padding(padding, padding)
+            .build();
+
+        let title_table_format = FormatBuilder::new()
+            .column_separator('║')
+            .borders('║')
+            .separators(&[LinePosition::Top], LineSeparator::new('═', '╦', '╔', '╗'))
+            .separators(
+                &[LinePosition::Intern],
+                LineSeparator::new('═', '╬', '╠', '╣'),
+            )
+            .padding(padding, padding)
+            .build();
```
What is the difference between these formats? They look almost the same :eyes:

AleX04Nov replied:
Yes, almost the same, except the title format does not have a bottom border.
I thought that it would be better to create a format for the title and then expand it with the bottom border for the main table. But style builder actually has method `build`, which I think - does not allow us to expand style. Also, `build` returns `TableFormat` instead of `TableFormatBuilder`.

But I think that I can save the builder into a mutable variable. And build those styles only when I would call `set_format` for my tables.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,575 @@
+    // Can't handle Colorized strings properly
+    fn create_table(
+        title: &str,
+        data: Vec<Vec<&String>>,
```
Very interesting argument type. Usually it would be written in one of the following ways:
```rust
data: Vec<Vec<String>> // When we want to took an ownership of the entire data structure to modify it and do something with the result (for example print). We took it by value so we can avoid allocations.

data: &[Vec<&str>] // We want just immutable reference to see the content and maybe print or seraillize
```

In our case, any of them will do the trick. And of course, your variant is ok, it just uses `&String` which requires two indirections instead of one, as in `&str`.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -0,0 +1,575 @@
+        if title.is_empty() {
+            Ok(main_table_str)
+        }
+        // Otherwise we need to pair them
```
Just a curiosity: have you tried to use recursive tables, (the example is at the end of [this](https://github.com/phsym/prettytable-rs#using-macros) section of pretty-table.rs)?  Maybe with some style magic it can be done easier. This is just an assumption :)

AleX04Nov replied:
No, I think I would be left with more headache and problems than with current logic.

AleX04Nov committed:
moved `encrypt_all_texts` method into `reinitialize_enc_dec`

Moved `encrypt_all_texts` method from `start` into `reinitialize_enc_dec` method.

AleX04Nov committed:
change argument type in `create_table`

Changed type of `data` argument in `create_table` method from `Vec<Vec<&String>>` to `&[Vec<&str>]`.

G1gg1L3s commented on cbc-repeated-iv/src/cbc_game.rs:
```diff
@@ -209,7 +209,7 @@ impl CBCGame {
-            vec![vec![&format!(
+            &[vec![&format!(
                 "\n{}\n{}\n{}\n{}\n{}\n\n",
                 &"Let's compare known CipherText with the leaked one:".to_string(),
```
Forgot to mention, that this `to_string()` is not needed

G1gg1L3s reviewed (approved):
Okay, looks good

AleX04Nov committed:
removed to_string from arguments of `create_table`

Removed to_string from arguments of `create_table`.

G1gg1L3s reviewed (approved):
Noice
